File: src/limanix/lima/template.py
"""Translate the validated Limanix configuration into a Lima template."""

import json
import os
import platform
from pathlib import Path, PurePosixPath

from limanix.config import Config
from limanix.domain import Architecture

# Published at github.com/nixos-lima/nixos-lima/releases/expanded_assets/v0.2.1.
_IMAGE_DIGESTS = {
    "aarch64": "ebdf8363bcb51542892963790c08ddfbffe45443ab19c5e70275c4f0c0aa6f11",
    "x86_64": "967da3baf4ea410e728c751ca9e0a617299b6809297e4e50e773cae4ce79197d",
}
_IMAGE_RELEASE = "https://github.com/nixos-lima/nixos-lima/releases/download/v0.2.1"


def uses_vz(arch: Architecture, host_arch: str | None = None) -> bool:
    """Use VZ for a native guest; a foreign architecture requires QEMU."""
    host = host_arch if host_arch is not None else platform.machine()
    return host in (arch.value, arch.lima_arch)


def render_lima(
    config: Config,
    managed_home: Path,
    runtime_dir: Path,
    *,
    host_arch: str | None = None,
    host_uid: int | None = None,
) -> str:
    """Return JSON, a YAML subset, with only explicitly managed host mounts.

    The dedicated management user applies NixOS configuration independently of
    the development user's sudo setting. ENV files live outside the staged flake.
    """
    arch = config.resources.arch.lima_arch
    native = uses_vz(config.resources.arch, host_arch)
    uid = os.getuid() if host_uid is None else host_uid
    mounts = [
        {
            "location": str(managed_home),
            "mountPoint": config.user.home,
            "writable": True,
        },
        {
            "location": str(runtime_dir),
            "mountPoint": "/mnt/limanix",
            "writable": False,
        },
        *(
            {
                "location": mount.source,
                "mountPoint": mount.target,
                "writable": mount.mode == "rw",
            }
            for mount in config.mounts
        ),
    ]
    mounts.sort(key=lambda mount: len(PurePosixPath(str(mount["mountPoint"])).parts))
    document = {
        "vmType": "vz" if native else "qemu",
        "arch": arch,
        "images": [
            {
                "location": f"{_IMAGE_RELEASE}/nixos-lima-v0.2.1-{arch}.qcow2",
                "arch": arch,
                "digest": f"sha256:{_IMAGE_DIGESTS[arch]}",
            }
        ],
        "cpus": config.resources.cpu,
        "memory": config.resources.mem.to_gib(),
        "disk": config.resources.disk.to_gib(),
        "user": {
            "name": "limanix-admin",
            "home": "/home/limanix-admin",
            "uid": 1001 if uid == 1000 else 1000,
            "passwordlessSudo": True,
        },
        "mountType": "virtiofs" if native else "9p",
        "mounts": mounts,
        "networks": [{"vzNAT": True}] if native else [{"lima": "shared"}],
        "ssh": {
            "loadDotSSHPubKeys": False,
            "forwardAgent": False,
        },
        "propagateProxyEnv": False,
        "portForwards": [
            {
                "guestIP": "0.0.0.0",
                "guestIPMustBeZero": False,
                "guestPortRange": [1, 65535],
                "proto": "any",
                "ignore": True,
            }
        ],
        "containerd": {"system": False, "user": False},
    }
    return json.dumps(document, indent=2) + "\n"

File: internal/lima/template.go
package lima

import (
	"encoding/json"
	"net"
	"sort"
	"strings"

	"github.com/lima-vm/lima/v2/pkg/limatype"
	"github.com/lima-vm/lima/v2/pkg/ptr"
	"github.com/mr-chelyshkin/limanix/internal/config"
	"github.com/mr-chelyshkin/limanix/internal/domain"
	"github.com/mr-chelyshkin/limanix/internal/nixos"
	"github.com/opencontainers/go-digest"
)

// Render uses Lima's schema types, emitting JSON (a YAML subset) without secret ENV values.
func Render(cfg config.Config, managedHome, runtimeDir string, hostArch domain.Architecture, hostUID int) ([]byte, error) {
	if _, err := domain.NewArchitecture(string(hostArch)); err != nil {
		return nil, err
	}

	arch, err := cfg.Resources.Arch.LimaArch()
	if err != nil {
		return nil, err
	}

	memory, err := cfg.Resources.Mem.GiB()
	if err != nil {
		return nil, err
	}

	disk, err := cfg.Resources.Disk.GiB()
	if err != nil {
		return nil, err
	}

	if hostUID <= 0 || managedHome == "" || runtimeDir == "" {
		return nil, ErrManagedMounts
	}

	image, err := nixos.BaseImage(cfg.Resources.Arch)
	if err != nil {
		return nil, err
	}

	document := machineTemplate(cfg.Resources.Arch, hostArch)
	document.Arch = &arch
	document.Images = []limatype.Image{{
		File: limatype.File{
			Location: image.Location,
			Arch:     arch,
			Digest:   digest.NewDigestFromEncoded(digest.SHA256, image.SHA256),
		},
	}}
	document.CPUs = ptr.Of(cfg.Resources.CPU)
	document.Memory = &memory
	document.Disk = &disk
	document.User = managementUser(hostUID)
	document.Mounts = templateMounts(cfg, managedHome, runtimeDir)

	encoded, err := json.MarshalIndent(document, "", "  ")
	if err != nil {
		return nil, err
	}

	return append(encoded, '\n'), nil
}

// usesVZ selects Apple's native hypervisor only for the hardware architecture.
func usesVZ(architecture, host domain.Architecture) bool {
	switch architecture {
	case domain.ARM64, domain.AMD64:
		return architecture == host
	default:
		return false
	}
}

func machineTemplate(architecture, host domain.Architecture) limatype.LimaYAML {
	document := limatype.LimaYAML{
		VMType:    ptr.Of(limatype.QEMU),
		MountType: ptr.Of(limatype.NINEP),
		Networks: []limatype.Network{
			{Lima: "shared"},
		},
		SSH: limatype.SSH{
			LoadDotSSHPubKeys: ptr.Of(false),
			ForwardAgent:      ptr.Of(false),
		},
		PropagateProxyEnv: ptr.Of(false),
		PortForwards: []limatype.PortForward{
			{
				GuestIP:           net.IPv4zero,
				GuestIPMustBeZero: ptr.Of(false),
				GuestPortRange:    [2]int{1, 65535},
				Proto:             limatype.ProtoAny,
				Ignore:            true,
			},
		},
		Containerd: limatype.Containerd{
			System: ptr.Of(false),
			User:   ptr.Of(false),
		},
	}

	if usesVZ(architecture, host) {
		document.VMType = ptr.Of(limatype.VZ)
		document.MountType = ptr.Of(limatype.VIRTIOFS)
		document.Networks = []limatype.Network{
			{VZNAT: ptr.Of(true)},
		}
	}

	return document
}

func managementUser(hostUID int) limatype.User {
	adminUID := uint32(1000)
	if hostUID == 1000 {
		adminUID = 1001
	}

	return limatype.User{
		Name:             ptr.Of("limanix-admin"),
		Home:             ptr.Of("/home/limanix-admin"),
		UID:              &adminUID,
		PasswordlessSudo: ptr.Of(true),
	}
}

func templateMounts(cfg config.Config, managedHome, runtimeDir string) []limatype.Mount {
	mounts := []limatype.Mount{
		{
			Location:   managedHome,
			MountPoint: ptr.Of(string(cfg.User.Home)),
			Writable:   ptr.Of(true),
		},
		{
			Location:   runtimeDir,
			MountPoint: ptr.Of("/mnt/limanix"),
			Writable:   ptr.Of(false),
		},
	}

	for _, mount := range cfg.Mounts {
		mounts = append(mounts, limatype.Mount{
			Location:   mount.Source,
			MountPoint: ptr.Of(string(mount.Target)),
			Writable:   ptr.Of(mount.Mode == "rw"),
		})
	}

	sort.SliceStable(mounts, func(i, j int) bool {
		return strings.Count(*mounts[i].MountPoint, "/") < strings.Count(*mounts[j].MountPoint, "/")
	})

	return mounts
}
